Library/Homebrew/os/mac/cache_store.rb

```diff
@@ -0,0 +1,207 @@
+  end
+
+  # Deletes data from the cache based on a condition defined in a concrete class
+  #
+  # @abstract
+  # @return [nil]
+  def flush_cache!
+    raise NotImplementedError
+  end
+
+  protected
+
+  # A class instance providing access to the `DBM` database object
+  #
+  # @return [DBM]
+  attr_reader :db
+end
+
+#
+# `LinkageStore` is a concrete class providing methods to fetch and mutate
+# linkage-specific data used by the `brew linkage` command
+#
+# If the cache hasn't changed, don't do extra processing in `LinkageChecker`.
+# Instead, just fetch the data stored in the cache
+#
+class LinkageStore < CacheStore
+  # Types of dylibs of the form (label -> array)
+  HASH_LINKAGE_TYPES = %w[brewed_dylibs reverse_links].freeze
+
+  # The keg name for the `LinkageChecker` class
+  #
+  # @return [String]
+  attr_reader :key
+
+  # Initializes new `LinkageStore` class
+  #
+  # @param  [String]        keg_name
+  # @param  [DatabaseCache] database_cache
+  # @return [nil]
+  def initialize(keg_name, database_cache)
+    @key = keg_name
+    super(database_cache)
+  end
+
+  # Inserts new values or updates existing cached values to persistent storage
+  # mechanism according to the type of data
+  #
+  # @param  [Hash] path_values
+  # @param  [Hash] hash_values
+  # @return [nil]
+  def update!(
+    path_values: {
+      "system_dylibs" => %w[], "variable_dylibs" => %w[], "broken_dylibs" => %w[],
+      "indirect_deps" => %w[], "undeclared_deps" => %w[], "unnecessary_deps" => %w[]
+    },
+    hash_values: {
+      "brewed_dylibs" => {}, "reverse_links" => {}
+    }
+  )
+    db[key] = {
+      "path_values" => format_path_values(path_values),
+      "hash_values" => format_hash_values(hash_values),
+    }
+  end
+
+  # Fetches cached values in persistent storage according to the type of data
+  # stored
+  #
+  # @param  [String] type
+  # @return [Any]
+  def fetch(type:)
+    if HASH_LINKAGE_TYPES.include?(type)
+      fetch_hash_values(type: type)
+    else
+      fetch_path_values(type: type)
+    end
+  end
+
+  # A condition for where to flush the cache
+  #
+  # @return [String]
+  def flush_cache!
+    db.delete(key)
+  end
+
+  private
+
+  # Fetches a subset of paths where the name = `key`
+  #
+  # @param  [String] type
+  # @return [Array[String]]
+  def fetch_path_values(type:)
+    return [] unless db.key?(key) && !db[key].nil?
+    string_to_hash(db[key])["path_values"][type]
+  end
+
+  # Fetches a subset of paths and labels where the name = `key`. Formats said
+  # paths/labels into `key => [value]` syntax expected by `LinkageChecker`
+  #
+  # @param  [String] type
+  # @return [Hash]
+  def fetch_hash_values(type:)
+    return {} unless db.key?(key) && !db[key].nil?
+    string_to_hash(db[key])["hash_values"][type]
+  end
+
+  # Parses `DBM` stored `String` into ruby `Hash`
+  #
+  # @param [String] string
+  # @return [Hash]
+  def string_to_hash(string)
+    JSON.parse(string.gsub("=>", ":"))
```

Comment:
Is `=>` escaped internally if that's e.g. within a string?

Reply:
Correct. DBM stores the ruby hash as a string. I converted the ruby string hash into a JSON string so that it could be parsed using the `JSON.parse` method.

Reply:
Where is `hash_to_string` or similar implemented? It feels weird that one way needs converted manually and the other does not.

Reply:
The `hash_to_string` is done implicitly by `dbm` when you insert the values into the database. Since `dbm` already does the conversion, I didn't think it was necessary to write a `hash_to_string` method, and store that result.

However, I could do a `hash_to_string` method before storing into `dbm`, where I could transform the ruby hash string into a JSON equivalent string. That way, `string_to_hash`'s logic would be simplified.

```ruby
  def string_to_hash(string)
    JSON.parse(string)
  end
```

I like this implementation better, so I think I'll change the logic around storing data into `dbm` so it converts it to a `JSON` compatible string first.

Reply:
> I like this implementation better, so I think I'll change the logic around storing data into dbm so it converts it to a JSON compatible string first.

Reply:
> I like this implementation better, so I think I'll change the logic around storing data into dbm so it converts it to a JSON compatible string first.

Yeh, that seems way better than doing any encoding/decoding ourselves.